In [1]:
from sklearn.datasets import load_digits
digits = load_digits()
print("Digits dataset loaded successfully.")

Digits dataset loaded successfully.


In [7]:
digits

{'data': array([[ 0.,  0.,  5., ...,  0.,  0.,  0.],
        [ 0.,  0.,  0., ..., 10.,  0.,  0.],
        [ 0.,  0.,  0., ..., 16.,  9.,  0.],
        ...,
        [ 0.,  0.,  1., ...,  6.,  0.,  0.],
        [ 0.,  0.,  2., ..., 12.,  0.,  0.],
        [ 0.,  0., 10., ..., 12.,  1.,  0.]]),
 'target': array([0, 1, 2, ..., 8, 9, 8]),
 'frame': None,
 'feature_names': ['pixel_0_0',
  'pixel_0_1',
  'pixel_0_2',
  'pixel_0_3',
  'pixel_0_4',
  'pixel_0_5',
  'pixel_0_6',
  'pixel_0_7',
  'pixel_1_0',
  'pixel_1_1',
  'pixel_1_2',
  'pixel_1_3',
  'pixel_1_4',
  'pixel_1_5',
  'pixel_1_6',
  'pixel_1_7',
  'pixel_2_0',
  'pixel_2_1',
  'pixel_2_2',
  'pixel_2_3',
  'pixel_2_4',
  'pixel_2_5',
  'pixel_2_6',
  'pixel_2_7',
  'pixel_3_0',
  'pixel_3_1',
  'pixel_3_2',
  'pixel_3_3',
  'pixel_3_4',
  'pixel_3_5',
  'pixel_3_6',
  'pixel_3_7',
  'pixel_4_0',
  'pixel_4_1',
  'pixel_4_2',
  'pixel_4_3',
  'pixel_4_4',
  'pixel_4_5',
  'pixel_4_6',
  'pixel_4_7',
  'pixel_5_0',
  'pixel_5_1',
 

In [2]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.2, random_state=42)

print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print(f"Shape of y_train: {y_train.shape}")
print(f"Shape of y_test: {y_test.shape}")

Shape of X_train: (1437, 64)
Shape of X_test: (360, 64)
Shape of y_train: (1437,)
Shape of y_test: (360,)


In [8]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

print(f"Mean of scaled X_train: {X_train.mean():.4f}")
print(f"Standard deviation of scaled X_train: {X_train.std():.4f}")
print("Feature normalization applied successfully.")

Mean of scaled X_train: -0.0000
Standard deviation of scaled X_train: 0.9763
Feature normalization applied successfully.


In [4]:
from sklearn.svm import SVC

# Instantiate the SVC classifier
svc_model = SVC(kernel='rbf', random_state=42)

# Train the SVC model
svc_model.fit(X_train, y_train)

print("SVC model trained successfully.")

SVC model trained successfully.


In [5]:
from sklearn.metrics import accuracy_score, confusion_matrix

# Make predictions on the test set
y_pred = svc_model.predict(X_test)

# Calculate accuracy score
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy Score: {accuracy:.4f}")

# Generate confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:\n", conf_matrix)

Accuracy Score: 0.9806

Confusion Matrix:
 [[33  0  0  0  0  0  0  0  0  0]
 [ 0 28  0  0  0  0  0  0  0  0]
 [ 0  0 33  0  0  0  0  0  0  0]
 [ 0  0  0 33  0  1  0  0  0  0]
 [ 0  0  0  0 46  0  0  0  0  0]
 [ 0  0  0  0  0 46  1  0  0  0]
 [ 0  0  0  0  0  0 35  0  0  0]
 [ 0  0  0  0  1  0  0 32  0  1]
 [ 0  0  0  0  1  0  0  0 29  0]
 [ 0  0  0  0  0  1  0  0  1 38]]


## Display Results and Discuss

### Subtask:
Display the classification accuracy, the confusion matrix, and provide a short discussion of the results.


### Discussion of Results

The Support Vector Classifier achieved an accuracy of **0.9806**, indicating that approximately 98.06% of the digits in the test set were classified correctly. This is a very high accuracy, suggesting strong performance by the model on this dataset.

The **Confusion Matrix** provides a detailed breakdown of the classification results:

```
[[33  0  0  0  0  0  0  0  0  0]
 [ 0 28  0  0  0  0  0  0  0  0]
 [ 0  0 33  0  0  0  0  0  0  0]
 [ 0  0  0 33  0  1  0  0  0  0]
 [ 0  0  0  0 46  0  0  0  0  0]
 [ 0  0  0  0  0 46  1  0  0  0]
 [ 0  0  0  0  0  0 35  0  0  0]
 [ 0  0  0  0  1  0  0 32  0  1]
 [ 0  0  0  0  1  0  0  0 29  0]
 [ 0  0  0  0  0  1  0  0  1 38]]
```

**Overall Performance:** The diagonal elements of the confusion matrix show the number of correctly classified instances for each digit. Most digits are classified with very high precision, as indicated by the large numbers on the diagonal and very few off-diagonal entries.

**Well-Classified Digits:** Digits such as '0', '1', '2', '4', and '6' appear to be almost perfectly classified with no misclassifications in the test set. For instance, all 33 instances of '0' were correctly identified.

**Misclassified Digits:** There are a few misclassifications, indicating areas where the model struggled:
- One instance of '3' was misclassified as '5'.
- One instance of '5' was misclassified as '6'.
- One instance of '7' was misclassified as '4', and another as '9'.
- One instance of '8' was misclassified as '4'.
- One instance of '9' was misclassified as '5', and another as '8'.

These misclassifications are minor given the high overall accuracy, but they suggest that these pairs of digits might share similar features that confuse the classifier. For example, '3' and '5' or '7', '4', and '9' might have similar stroke patterns or shapes. Further analysis could involve examining these specific misclassified samples to understand the reasons for the errors, potentially leading to further model improvements or feature engineering.

## Final Task

### Subtask:
Summarize the machine learning pipeline implementation, including the chosen dataset, model, and the performance metrics obtained.


## Summary:

### Q&A
The machine learning pipeline implemented uses the Digits dataset, trains a Support Vector Classifier (SVC) with an RBF kernel, and evaluates its performance using accuracy and a confusion matrix. The model achieved an accuracy of 0.9806.

### Data Analysis Key Findings
*   **Dataset**: The Digits dataset from scikit-learn, consisting of 1797 samples with 64 features each, was successfully loaded.
*   **Data Preparation**: The dataset was split into training and testing sets with an 80/20 ratio, resulting in 1437 training samples and 360 testing samples. Feature normalization was applied using `StandardScaler`, ensuring the training features had a mean of approximately 0 and a standard deviation of 1.
*   **Model**: A Support Vector Classifier (SVC) with a Radial Basis Function (RBF) kernel and `random_state=42` was instantiated and trained on the normalized data.
*   **Performance Metrics**:
    *   The model achieved a high accuracy score of **0.9806** on the test set.
    *   The confusion matrix revealed that most digits were classified correctly. Digits '0', '1', '2', '4', and '6' were almost perfectly classified with no misclassifications.
    *   Minor misclassifications were observed for digits '3' (1 instance), '5' (1 instance), '7' (2 instances), '8' (1 instance), and '9' (2 instances). For example, one '3' was misclassified as '5', and one '7' was misclassified as '4'.

### Insights or Next Steps
*   The high accuracy demonstrates that the SVC model, combined with feature normalization, is highly effective for digit classification on the Digits dataset.
*   Further analysis could focus on the specific misclassified digits identified in the confusion matrix to understand why the model struggled with them, potentially leading to targeted feature engineering or exploration of more complex models if necessary.
